Support TSQLs Information_schema.schemata Changes

bsatarupa opened the pull request:
### Description
Implemented information_schema_tsql.schemata view which returns one row for each schema in the current database containing catalog_name, schema_name, schema_owner, DEFAULT_CHARACTER_SET_CATALOG, DEFAULT_CHARACTER_SET_SCHEMA and DEFAULT_CHARACTER_SET_NAME.

Pending task for BBF: 
We need to create mapping of collation name to char-set name; until then returning null for DEFAULT_CHARACTER_SET_NAME
Could not test with explicit authorization role in most version upgrade version, as 'create role/create user' was not supported in BBF back then.

Added testcases for information_schema.schemata in JDBC.

Task: BABEL-2996
Signed-off-by: Satarupa Biswas <satarupb@amazon.com>

### Test scenarios covered

 o Use case based: Done
          - Shown schema detail for all schemas under any database.
          - Tested schemata in cross reference databases; created schema inside a newly created database, and showed that the same can't be accessed from master databases or other databases.
          - Tested with schemas created by explicit user
o Boundary conditions: N/A
o Arbitrary inputs : Done
o Null, Negative test scenarios: N/A
o Minor version upgrade tests, Major version upgrade tests : Done on 14.5 and above, not added to all version since back then CREATE USER/CREATE ROLE was not supported in Babelfish.
o Performance tests, Tooling impact, Client tests: N/A

### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

Deepesh125 commented:
Please remove BABEL-XXX from PR title. Plus, add all details of test cases added with this PR in PR description. And please fix github actions which are failing.

Deepesh125 commented on contrib/babelfishpg_tsql/sql/information_schema_tsql.sql:
```diff
@@ -716,4 +716,35 @@ CREATE OR REPLACE VIEW information_schema_tsql.routines AS
+		-- Until then return null for DEFAULT_CHARACTER_SET_NAME
+		CAST(null AS sys.sysname) AS "DEFAULT_CHARACTER_SET_NAME"
+	FROM (sys.pg_namespace_ext nc JOIN pg_catalog.pg_roles r on r.oid = nc.nspowner)
+		LEFT JOIN sys.babelfish_namespace_ext ext on nc.nspname = ext.nspname
```
information_schema.schemata should return details of system schemas also (guest, information_schema, sys, etc.)

bsatarupa replied:
done

Deepesh125 commented on test/JDBC/expected/BABEL_SCHEMATA-vu-prepare.out:
```diff
@@ -0,0 +1,22 @@
+go
+
+create schema schema_21;
+go
```
switch back to master database.

bsatarupa replied:
done

Deepesh125 replied:
use master is still missing here

Deepesh125 commented on contrib/babelfishpg_tsql/sql/information_schema_tsql.sql:
```diff
@@ -716,4 +716,35 @@ CREATE OR REPLACE VIEW information_schema_tsql.routines AS
+			'aws_s3', 'aws_lambda', 'pglogical', 'dbo')
+			THEN CAST(ext.orig_name AS sys.nvarchar(128))
+		WHEN r.rolname LIKE CONCAT(nc.dbname,'%')
+			THEN CAST(RIGHT(r.rolname, LENGTH(r.rolname) - LENGTH(nc.dbname) - 1) AS sys.nvarchar(128))
```
split_part would be better option to use here.

bsatarupa replied:
split_part() is to split string based on a given delimiter and its position in the string; in these case where rolname is a concat of dbname and schema_owner; we can't determine the exact position of delimiter '-' where there can be multiple occurrences of '-' in both dbname and schema_owner name.
Instead, r.rolname LIKE CONCAT(nc.dbname,'%') is checking if rolname has a prefix with dbname, if yes, it is extracting {LENGTH(r.rolname) - LENGTH(nc.dbname) - 1} length of data from right side of rolname as schema_owher name.

Deepesh125 commented on contrib/babelfishpg_tsql/sql/information_schema_tsql.sql:
```diff
@@ -716,4 +716,35 @@ CREATE OR REPLACE VIEW information_schema_tsql.routines AS
+		-- For user-defined schemas having default owner, schema-owner will be dbo
+		-- For user-defined schemas with explicit owners, rolname contains dbname followed
+		-- by owner name, so need to extract the owner name from rolname always. 
+		CASE WHEN CAST(ext.orig_name AS sys.nvarchar(128)) IN
+			('sys', 'information_schema_tsql', 'public',
+			'pg_catalog', 'pg_toast', 'information_schema', 'aws_commons',
+			'aws_s3', 'aws_lambda', 'pglogical', 'dbo')
+			THEN CAST(ext.orig_name AS sys.nvarchar(128))
```
Is there any existing function/mechanism available which is meant for this only? Check with Harsh or Shalini.

bsatarupa replied:
There is one but that is C function: is_shared_schema(); have added code to expose this C function from sql

Deepesh125 commented on test/JDBC/expected/BABEL_SCHEMATA-vu-verify.out:
```diff
@@ -0,0 +1,26 @@
+db_20#!#dbo#!#dbo#!#<NULL>#!#<NULL>#!#<NULL>
+db_20#!#schema_21#!#dbo#!#<NULL>#!#<NULL>#!#<NULL>
+~~END~~
+
```
switch back to master db

bsatarupa replied:
done

Deepesh125 commented on test/JDBC/input/BABEL_SCHEMATA.mix:
```diff
@@ -0,0 +1,39 @@
+-- [BABEL-2996]Testing Cross reference schema
```
is this file copy of BABEL_SCHEMATA-vu-prepare.sql + BABEL_SCHEMATA-vu-verify.sql

bsatarupa replied:
yes

Deepesh125 commented on contrib/babelfishpg_tsql/sql/information_schema_tsql.sql:
```diff
@@ -716,4 +716,35 @@ CREATE OR REPLACE VIEW information_schema_tsql.routines AS
+		LEFT JOIN sys.babelfish_namespace_ext ext on nc.nspname = ext.nspname
+	WHERE ext.dbid = cast(sys.db_id() as oid)
+		AND (pg_has_role(nc.nspowner, 'USAGE')
+		OR has_schema_privilege(nc.oid, 'CREATE, USAGE')) ORDER BY ext.orig_name;
```
Order by clause is not required here. Instead use it test case to make to order deterministic.

bsatarupa replied:
Removing order by is causing non-deterministic sequence failure even when upgrade script has order by clause:

 varchar#!#nvarchar#!#nvarchar#!#varchar#!#varchar#!#varchar
 master#!#babelfish_namespace_ext_vu_prepare_sch1#!#dbo#!#<NULL>#!#<NULL>#!#<NULL>
 master#!#babelfish_namespace_ext_vu_prepare_sch2#!#dbo#!#<NULL>#!#<NULL>#!#<NULL>
-master#!#dbo#!#dbo#!#<NULL>#!#<NULL>#!#<NULL>
+master#!#objectproperty_vu_prepare_ownerid_schema#!#dbo#!#<NULL>#!#<NULL>#!#<NULL>
 master#!#schema_resolution_func_vu_prepare_s1#!#dbo#!#<NULL>#!#<NULL>#!#<NULL>
-master#!#schema_resolution_func_vu_prepare_s2#!#dbo#!#<NULL>#!#<NULL>#!#<NULL>
 master#!#schema_resolution_proc_sch1#!#dbo#!#<NULL>#!#<NULL>#!#<NULL>
-master#!#schema_resolution_proc_sch2#!#dbo#!#<NULL>#!#<NULL>#!#<NULL>
 master#!#schema_resolution_trigger_s1#!#dbo#!#<NULL>#!#<NULL>#!#<NULL>
 master#!#sys_all_parameters_schema_vu_prepare#!#dbo#!#<NULL>#!#<NULL>#!#<NULL>
+master#!#sys_events_vu_prepare_schema1#!#dbo#!#<NULL>#!#<NULL>#!#<NULL>
 master#!#sys_schemas_dep_vu_prepare_test1#!#dbo#!#<NULL>#!#<NULL>#!#<NULL>
 master#!#sys_schemas_vu_prepare_test1#!#dbo#!#<NULL>#!#<NULL>#!#<NULL>
 master#!#sys_schemas_vu_prepare_test2#!#dbo#!#<NULL>#!#<NULL>#!#<NULL>
+master#!#BABEL_3117_schemaC#!#dbo#!#<NULL>#!#<NULL>#!#<NULL>
+master#!#dbo#!#dbo#!#<NULL>#!#<NULL>#!#<NULL>
+master#!#objectpropertyex_ownerid_schema#!#dbo#!#<NULL>#!#<NULL>#!#<NULL>
+master#!#schema_resolution_func_vu_prepare_s2#!#dbo#!#<NULL>#!#<NULL>#!#<NULL>
+master#!#schema_resolution_proc_sch2#!#dbo#!#<NULL>#!#<NULL>#!#<NULL>
+master#!#sys_trigger_events_vu_prepare_schema1#!#dbo#!#<NULL>#!#<NULL>#!#<NULL>

surendravishnoi commented on test/JDBC/upgrade/latest/schedule:
```diff
@@ -327,3 +327,4 @@ BABEL_GRANT_CONNECT
 BABEL-sp_helpdb
 BABEL-2795
 babelfish_integrity_checker
+BABEL_SCHEMATA
```
latest to latest only does MVU. This also need minor version test, add to 14.5 to latest or similar

bsatarupa replied:
Alright, added on 14.5 and above, not added to all version since back then CREATE USER/CREATE ROLE was not supported in Babelfish.

Deepesh125 commented on contrib/babelfishpg_tsql/sql/information_schema_tsql.sql:
```diff
@@ -716,4 +716,35 @@ CREATE OR REPLACE VIEW information_schema_tsql.routines AS
+	CAST(null AS sys.sysname) AS "DEFAULT_CHARACTER_SET_NAME"
+	FROM ((pg_catalog.pg_namespace np LEFT JOIN sys.pg_namespace_ext nc on np.nspname = nc.nspname)
+		LEFT JOIN pg_catalog.pg_roles r on r.oid = nc.nspowner) LEFT JOIN sys.babelfish_namespace_ext ext on nc.nspname = ext.nspname
+	WHERE (ext.dbid = cast(sys.db_id() as oid) OR np.nspowner = 10) AND
```
what does this np.nspowner = 10 represents?

bsatarupa replied:
np.nspowner = 10 means rdsadmin
It is used to get all schemas under pg_catalog.pg_namespace (sys, information_schema, public etc. etc.)

Deepesh125 commented on contrib/babelfishpg_tsql/sql/information_schema_tsql.sql:
```diff
@@ -716,4 +716,35 @@ CREATE OR REPLACE VIEW information_schema_tsql.routines AS
+	FROM ((pg_catalog.pg_namespace np LEFT JOIN sys.pg_namespace_ext nc on np.nspname = nc.nspname)
+		LEFT JOIN pg_catalog.pg_roles r on r.oid = nc.nspowner) LEFT JOIN sys.babelfish_namespace_ext ext on nc.nspname = ext.nspname
+	WHERE (ext.dbid = cast(sys.db_id() as oid) OR np.nspowner = 10) AND
+	      (pg_has_role(nc.nspowner, 'USAGE') OR has_schema_privilege(nc.oid, 'CREATE, USAGE')
+	      OR pg_has_role(np.nspowner, 'USAGE') OR has_schema_privilege(np.oid, 'CREATE, USAGE'))
```
This two conditions are essentially same. We can get rid of one of the line.

bsatarupa replied:
okay

Deepesh125 commented on contrib/babelfishpg_tsql/sql/sys_function_helpers.sql:
```diff
@@ -10293,3 +10293,8 @@ BEGIN
+CREATE OR REPLACE FUNCTION sys.bbf_is_shared_schema(IN schemaname TEXT)
+RETURNS BOOL
+AS 'babelfishpg_tsql', 'is_shared_schema_wrapper'
+LANGUAGE C STABLE STRICT;
```
We can make this function immutable

bsatarupa replied:
done

HarshLunagariya commented on contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--2.2.0--2.3.0.sql:
```diff
@@ -4229,3 +4229,37 @@ BEGIN
+AS 'babelfishpg_tsql', 'is_shared_schema_wrapper'
+LANGUAGE C IMMUTABLE STRICT;
+
+CREATE OR REPLACE VIEW information_schema_tsql.schemata AS
+	SELECT CAST(sys.db_name() AS sys.sysname) AS "CATALOG_NAME",
+	CAST(CASE WHEN np.nspname LIKE CONCAT(sys.db_name(),'%') THEN RIGHT(np.nspname, LENGTH(np.nspname) - LENGTH(sys.db_name()) - 1)
+	     ELSE np.nspname END AS sys.nvarchar(128)) AS "SCHEMA_NAME",
+	-- For system-defined schemas, schema-owner name will be same as schema_name
+	-- For user-defined schemas having default owner, schema-owner will be dbo
+	-- For user-defined schemas with explicit owners, rolname contains dbname followed
+	-- by owner name, so need to extract the owner name from rolname always.
+	CAST(CASE WHEN sys.bbf_is_shared_schema(np.nspname) = TRUE THEN np.nspname
+		  WHEN r.rolname LIKE CONCAT(sys.db_name(),'%') THEN
+			CASE WHEN RIGHT(r.rolname, LENGTH(r.rolname) - LENGTH(sys.db_name()) - 1) = 'db_owner' THEN 'dbo'
+			     ELSE RIGHT(r.rolname, LENGTH(r.rolname) - LENGTH(sys.db_name()) - 1) END ELSE 'dbo' END
+			AS sys.nvarchar(128)) AS "SCHEMA_OWNER",
+	CAST(null AS sys.varchar(6)) AS "DEFAULT_CHARACTER_SET_CATALOG",
+	CAST(null AS sys.varchar(3)) AS "DEFAULT_CHARACTER_SET_SCHEMA",
+	-- TODO: We need to first create mapping of collation name to char-set name;
+	-- Until then return null for DEFAULT_CHARACTER_SET_NAME
+	CAST(null AS sys.sysname) AS "DEFAULT_CHARACTER_SET_NAME"
+	FROM ((pg_catalog.pg_namespace np LEFT JOIN sys.pg_namespace_ext nc on np.nspname = nc.nspname)
+		LEFT JOIN pg_catalog.pg_roles r on r.oid = nc.nspowner) LEFT JOIN sys.babelfish_namespace_ext ext on nc.nspname = ext.nspname
+	WHERE (ext.dbid = cast(sys.db_id() as oid) OR np.nspname in ('sys', 'information_schema_tsql')) AND
+	      (pg_has_role(np.nspowner, 'USAGE') OR has_schema_privilege(np.oid, 'CREATE, USAGE'))
+	ORDER BY nc.nspname, np.nspname;
```
We can avoid multiple sys.db_name() calls here, instead we can use pg_namespace_ext.dbname.

bsatarupa replied:
pg_namespace_ext.dbname is different;  it is giving null for most of the entries and for pg schemas giving pg_database name; here sys.db_name() is replicating sql server behaviour.

surendravishnoi commented on contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--2.2.0--2.3.0.sql:
```diff
@@ -4229,3 +4229,37 @@ BEGIN
             RETURN NULL;
 END; $BODY$
 LANGUAGE plpgsql;
+
```
if this is for ocelot, file should change to 2.3 to 2.4

bsatarupa replied:
done

surendravishnoi commented on test/JDBC/input/BABEL_SCHEMATA-vu-prepare.sql:
```diff
@@ -0,0 +1,22 @@
+-- [BABEL-2996]Testing Cross reference schema
+create database db_10;
```
this looks like very simple test case for supporting a new view, please sync with Harsh/Shalini to find more variation to cover.

bsatarupa replied:
Added check that after revoking the usage permission/ownership from user whether it is still showing in catalog or not.

surendravishnoi commented on test/python/expected/upgrade_validation/expected_dependency.out:
```diff
@@ -223,6 +223,7 @@ Function sys.babelfish_waitfor_delay(text)
 Function sys.babelfish_waitfor_delay(timestamp without time zone)
 Function sys.bbf_binary_cmp(sys.bbf_binary,sys.bbf_binary)
 Function sys.bbf_get_current_physical_schema_name(text)
+Function sys.bbf_is_shared_schema(text)
```
this indicates that we have not added dependency test for this view

bsatarupa replied:
Fixed

surendravishnoi commented on contrib/babelfishpg_tsql/sql/information_schema_tsql.sql:
```diff
@@ -716,4 +716,33 @@ CREATE OR REPLACE VIEW information_schema_tsql.routines AS
+/*
+ * SCHEMATA view
+ */
+CREATE OR REPLACE VIEW information_schema_tsql.schemata AS
```
for testing, this can have impact on tooling and should be validated. Also, we should compare with SQL server to make sure that view has reasonable perf for large number of objects.

Commit:
Support TSQLs Information_schema.schemata Changes

Implemented information_schema_tsql.schemata view which returns one row for each schema in the current database containing catalog_name, schema_name, schema_owner, DEFAULT_CHARACTER_SET_CATALOG, DEFAULT_CHARACTER_SET_SCHEMA and DEFAULT_CHARACTER_SET_NAME.

Pending task for BBF:
We need to create mapping of collation name to char-set name; until then returning null for DEFAULT_CHARACTER_SET_NAME
Could not test with explicit authorization role in most version upgrade version, as 'create role/create user' was not supported in BBF back then.

Added testcases for information_schema.schemata in JDBC.

Task: BABEL-2996
Signed-off-by: Satarupa Biswas satarupb@amazon.com

HarshLunagariya committed:
Update expected_dependency output

Signed-off-by: Harsh Lunagariya <lunharsh@amazon.com>

HarshLunagariya committed:
Merge branch 'BABEL_3_X_DEV' into jira-babel-2996

HarshLunagariya committed:
Merge branch 'BABEL_3_X_DEV' into jira-babel-2996

Deepesh125 commented on contrib/babelfishpg_tsql/sql/information_schema_tsql.sql:
```diff
@@ -750,4 +750,33 @@ CREATE OR REPLACE VIEW information_schema_tsql.SEQUENCES AS
+ * SCHEMATA view
+ */
+CREATE OR REPLACE VIEW information_schema_tsql.schemata AS
+	SELECT CAST(sys.db_name() AS sys.sysname) AS "CATALOG_NAME",
```
not sure if calling sys.db_name repeatedly good for performance as it is being used at multiple places. Should try to see if we can have an extra join with db_name.

KushaalShroff replied:
I am not a big fan to add this here as well but this should not be a perfomance issue since optimizer will rewrite it as a literal in the plan.

Deepesh125 commented on contrib/babelfishpg_tsql/sql/information_schema_tsql.sql:
```diff
@@ -750,4 +750,33 @@ CREATE OR REPLACE VIEW information_schema_tsql.SEQUENCES AS
+	-- TODO: We need to first create mapping of collation name to char-set name;
+	-- Until then return null for DEFAULT_CHARACTER_SET_NAME
+	CAST(null AS sys.sysname) AS "DEFAULT_CHARACTER_SET_NAME"
+	FROM ((pg_catalog.pg_namespace np LEFT JOIN sys.pg_namespace_ext nc on np.nspname = nc.nspname)
+		LEFT JOIN pg_catalog.pg_roles r on r.oid = nc.nspowner) LEFT JOIN sys.babelfish_namespace_ext ext on nc.nspname = ext.nspname
+	WHERE (ext.dbid = cast(sys.db_id() as oid) OR np.nspname in ('sys', 'information_schema_tsql')) AND
+	      (pg_has_role(np.nspowner, 'USAGE') OR has_schema_privilege(np.oid, 'CREATE, USAGE'))
+	ORDER BY nc.nspname, np.nspname;
```
Check with DBE if we can optimise this query further.

KushaalShroff replied:
Its efficient and reliant on pg_catalog tables and looks good to me. Have asked Harsh to provide some perf analysis to prove it.

Deepesh125 commented on test/python/expected/upgrade_validation/expected_dependency.out:
```diff
@@ -954,6 +955,7 @@ View information_schema_tsql.column_domain_usage
 View information_schema_tsql.constraint_column_usage
 View information_schema_tsql.domains
 View information_schema_tsql.routines
+View information_schema_tsql.schemata
```
we need to add dependency test cases.

HarshLunagariya replied:
Added

HarshLunagariya committed:
Merge branch 'BABEL_3_X_DEV' into jira-babel-2996

HarshLunagariya committed:
Update expected_dependency.out

HarshLunagariya committed:
Add dependency tests

Signed-off-by: Harsh Lunagariya <lunharsh@amazon.com>

HarshLunagariya committed:
Update schedule

HarshLunagariya committed:
Fix BABEL_SCHEMATA test

Signed-off-by: Harsh Lunagariya <lunharsh@amazon.com>

HarshLunagariya committed:
Rename dep test

Signed-off-by: Harsh Lunagariya <lunharsh@amazon.com>

HarshLunagariya committed:
Update schedule

HarshLunagariya committed:
Add BABEL_SCHEMATA to remaining schedule files

Signed-off-by: Harsh Lunagariya <lunharsh@amazon.com>

HarshLunagariya committed:
Merge branch 'BABEL_3_X_DEV' into jira-babel-2996

HarshLunagariya committed:
Update schedule

HarshLunagariya committed:
Update jdbc_schedule

HarshLunagariya committed:
Update 14_7 schedule

KushaalShroff reviewed (dismissed):
Definition LGTM

HarshLunagariya committed:
Merge branch 'babelfish-for-postgresql:BABEL_3_X_DEV' into jira-babel-2996

HarshLunagariya committed:
Merge branch 'BABEL_3_X_DEV' into jira-babel-2996

HarshLunagariya committed:
Update BABEL_SCHEMATA test to not show unrelated schemas from other tests

Signed-off-by: Harsh Lunagariya <lunharsh@amazon.com>

HarshLunagariya committed:
Change object names in BABEL_SCHEMATA test

Signed-off-by: Harsh Lunagariya <lunharsh@amazon.com>
